materials/variation_with_state.py

```diff
@@ -0,0 +1,254 @@
+        domain_str_list = []
+        domain = self.get_state_domain()    #pylint: disable=assignment-from-no-return
+        for name in self.state_vars:
+            domain_str_list.append(
+                '{:.4g} to {:.4g} {:s}'.format(
+                    domain[name][0], domain[name][1], self.state_vars_units[name]))
+        domain_str = ', '.join(domain_str_list)
+        return 'Variation with {:s} over {:s}, represented as a {:s}. [Data from {:s}]'.format(
+            state_var_str, domain_str, self.representation, self.reference)
+
+
+class VariationWithStateTable(VariationWithState):
+    """A material property's variation with state, represented as a table."""
+    def __init__(self, state_vars, state_vars_units, value_type, reference,
+                 interp_points, interp_values, state_vars_interp_scales):
+        VariationWithState.__init__(self, 'table', state_vars, state_vars_units, value_type, reference)
+        self._interp_points = interp_points
+        self._interp_values = interp_values
+        self._state_vars_interp_scales = state_vars_interp_scales
+
+    def query_value(self, state, method='linear', fill_value=np.nan, rescale=True):
+        """Query the value of the property at a particular state.
+
+        Arguments:
+            state (dict): The state at which to query the values. It must have
+                a key for each variable name in `self.state_vars`. `state[s1]`
+                specifies the query point for state variable `s1`. The query point
+                for each state may be an array or a scalar. e.g.
+                    `state={'s1': 0, 's2': 1}`
+                    `state={'s1': 0, 's2': [1, 2, 3]}`
+                    and
+                    `state={'s1': [5, 6, 7], 's2': [1, 2, 3]}`
+                are all valid.
+        `method`, `fill_value`, and `rescale` are passed through to `scipy.interpolate.griddata`.
+
+        Returns:
+            scalar or array: value(s) of the property at the provided state(s).
+        """
+        # Check that all the state variables have been provided in `state`.
+        for var_name in self.state_vars:
+            if var_name not in state.keys():
+                raise ValueError('{:s} not provided for query'.format(var_name))
+
+        # Copy the state argument - we might need to mutate it.
+        state = copy.deepcopy(state)
+
+        # Form an array of points at which to query the interpolation.
+        is_state_scalar = True
+        if len(self.state_vars) == 1:
+            query_points = np.array(state[self.state_vars[0]])
+            if hasattr(query_points, '__len__'):
+                is_state_scalar = False
+            if self._state_vars_interp_scales[0] == 'log':
+                query_points = np.log(query_points)
+        elif len(self.state_vars) == 2:
+            # Handle cases with 1 array and 1 scalar state
+            if hasattr(state[self.state_vars[0]], '__len__') \
```

Comment:
Consider if you can come up with a strategy that doesn't require this casing. 
One could:
- Put more onus on the user to provide valid input (And throw on invalid input)
- Reorganize so that it's only called with one type of input

Checking explicitly for the existence of methods can be brittle -- what would guido say!?

Reply:
I replaced `hasattr(x, '__len__')` with `np.ndim(x) > 0` in 037d612e6507212ccf1232983d3ab1e646663d51

When I was writing the demos I found it was convenient to be able to use any of {(scalar, scalar), (scalar, array), (array, array)} in a query.

Another contribution to the messiness of this casing is checking whether the lookup table is 1D or 2D (`len(self.state_vars) == 2`). I could split `VariationWithStateTable` into two subclasses, one to handle 1D and one to handle 2D. This would be about an hour of work; do you think it would significantly improve readability?

Reply:
I think it would be noble, and might free you from future bugs, but it might make sense to get the project to an MVP state and then pursue some small refactors